lib/elixir/lib/calendar/date.ex

```diff
@@ -757,6 +769,75 @@ defmodule Date do
     end
   end
 
+  @doc """
+  Shifts a date by given units. Raises an `ArgumentError` if the
+  given date is not a valid Calendar.ISO date.
+
+  Available shift units are: `:year, :month, :week, :day` and the shift
+  is always applied in that order.
+
+  `Date.shift/2` defines a week as 7 days and a year as 12 months.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - if the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Date.shift(~D[2016-01-03], month: 2)
+      ~D[2016-03-03]
+      iex> Date.shift(~D[2016-02-29], month: 1)
+      ~D[2016-03-29]
+      iex> Date.shift(~D[2016-01-31], month: 1)
+      ~D[2016-02-29]
+      iex> Date.shift(~D[2016-01-31], year: 4, day: 1)
+      ~D[2020-02-01]
+
+  """
+  # @doc since: "1.5.0"
+  @spec shift(Calendar.date(), [shift_unit()]) :: t
+  def shift(%{calendar: Calendar.ISO} = date, shift_units) do
+    shift_units = Keyword.validate!(shift_units, year: 0, month: 0, week: 0, day: 0)
+
+    ordered_units = [
+      year: shift_units[:year],
+      month: shift_units[:month],
+      week: shift_units[:week],
+      day: shift_units[:day]
+    ]
+
+    Enum.reduce(ordered_units, date, fn
+      {_opt, 0}, new_date -> new_date
+      {:year, value}, new_date -> shift_months(new_date, value * 12)
+      {:month, value}, new_date -> shift_months(new_date, value)
+      {:week, value}, new_date -> Date.add(new_date, value * 7)
+      {:day, value}, new_date -> Date.add(new_date, value)
+    end)
+  end
+
+  def shift(_date, _opts) do
+    raise ArgumentError, "date invalid or not supported"
+  end
+
+  @doc false
+  defp shift_months(date, 0), do: date
+
+  defp shift_months(%Date{year: year, month: month, day: day}, months) do
+    total_months = year * 12 + month + months - 1
+    new_year = Integer.floor_div(total_months, 12)
+
+    new_month =
+      case rem(total_months, 12) + 1 do
+        0 -> 12
+        new_month -> new_month
+      end
+
+    last_day_of_month = :calendar.last_day_of_the_month(abs(new_year), new_month)
+    new_day = min(day, last_day_of_month)
```

Comment:
In [ex_cldr_calendars](https;//github.com/elixir-cldr/cldr_calendars) I decided to use a `:coerce` boolean option to determine whether to force the date to be valid (using `min/2` as you have here) or not. This only matters for `:year` and `:month` units. I've never been happy with the name `:coerce`  but the principle still might be useful.

Reply:
I think `:round` would be a good name. With `round: :nowhere` perhaps meaning it is invalid.

Reply:
> I think `:round` would be a good name. With `round: :nowhere` perhaps meaning it is invalid.

Yes, thats a much  better name.